### DISCUSSION TITLE
Trying to build robocode tank royale (gradle errors)

### DISCUSSION THREAD

**Discussion by dEgolf** [robocode-dev/tank-royale#6] [Q&A] [answered]
Hi again! 
I'm trying to build robocode tank royal using gradle. (My motivation for doing this is that I'd like to understand better exactly what API commands are doing at a low level, so I can better control my robot). However, the build fails for me with this information:

```* Where:
* What went wrong:
Cannot get non-null property 'robocodeDevOssrhUsername' on project ':bot-api:java' as it does not exist
```

I'm using Windows 10, and the following versions of some relevant software:
```
$ java --version
openjdk 11.0.14.1 2022-02-08
OpenJDK Runtime Environment Temurin-11.0.14.1+1 (build 11.0.14.1+1)
OpenJDK 64-Bit Server VM Temurin-11.0.14.1+1 (build 11.0.14.1+1, mixed mode)

$ dotnet --version
6.0.201

$ gradle -v

------------------------------------------------------------
Gradle 7.4.1
------------------------------------------------------------

Build time:   2022-03-09 15:04:47 UTC
Revision:     36dc52588e09b4b72f2010bc07599e0ee0434e2e

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          11.0.14.1 (Eclipse Adoptium 11.0.14.1+1)
OS:           Windows 10 10.0 amd64
```

Any suggestions are appreciated.

Edit:
My  current guess is that a file is supposed to be created and then saved on OSSRH, which requires credentials. If this is the problem, might it be possible to configure the build to save the file locally instead of on OSSRH?

Edit 2:
I tried commenting out things that seemed to be causing errors. I then ended up with this error:
```
* What went wrong:
Could not determine the dependencies of task ':server:shadowJar'.
> Could not resolve all dependencies for configuration ':server:runtimeClasspath'.
   > Could not find dev.robocode.tankroyale:robocode-tankroyale-schema:0.10.0.
     Searched in the following locations:
       - file:/C:/Users/David/.m2/repository/dev/robocode/tankroyale/robocode-tankroyale-schema/0.10.0/robocode-tankroyale-schema-0.10.0.pom
       - https://repo.maven.apache.org/maven2/dev/robocode/tankroyale/robocode-tankroyale-schema/0.10.0/robocode-tankroyale-schema-0.10.0.pom
     Required by:
         project :server
```
I notice that the URL listed in the error message doesn't exist, although this related one does: "https://repo.maven.apache.org/maven2/dev/robocode/tankroyale/robocode-tankroyale-bot-api/0.10.0/robocode-tankroyale-bot-api-0.10.0.pom". Maybe it's looking in the wrong spot?

**Comment by flemming-n-larsen** [comment, marked as answer]
You are right that it was used for the Sonarype OSSRH. I have commented this out for now, so you can get a new version by doing a `git pull`  from the master branch.

The error you see with the robocode-tankroyale-schema is because you are missing this artifact in your local maven repository.

I might have a look into making this a bit smarter. But then again. The schemas are not changed that often.

Steps:
- Go to the /schema/java
- Do a `gradle build`
- Do a `gradle publishToMavenLocal` to publish the artifact your your local repository

Now you should be able to build the Bot API for Java. :)

**Comment by flemming-n-larsen** [comment]
Sorry, the last line needs to be:
```
gradle publishToMavenLocal
```

**Reply by dEgolf** [reply]
Great! I think I got it building. 
Here's what I did:
- As far as I could tell, the Sonarype OSSRH stuff wasn't commented out, even after doing a `git pull`. 
- So, I commented out both occurrences of this block of code in `bot-api\java\build.gradle.kts`:
```
credentials {
    username = robocodeDevOssrhUsername
    password = robocodeDevOssrhPassword
}
```
- I also ended up commenting out this code in `gui-app\build.gradle.kts` and `schema\java\build.gradle.kts`:
```
signing {
    sign(publishing.publications["mavenJava"])
}
```
- I commented out `apply(from = "../../groovy.gradle")` and the whole section under `register("upload")` in `sample-bots\csharp\build.gradle.kts` and in `sample-bots\java\build.gradle.kts`.
- I also commented out `include("buildDocs")` in `settings.gradle.kts`.

- I was then able to successfully run `gradle build` and `gradle publishToMavenLocal` in /schema/java.

I was confused as to where the resulting files were placed once the build succeeded, but it turns out they were saved in the Maven local repository at `C:\Users\David\.m2\repository\dev\robocode\tankroyale`.

**Reply by flemming-n-larsen** [reply]
I have updated the build file. For some reason. the push did not succeed, and I have an issue updating my previous comments due to some temporary GitHub issue. But it is okay now.

I see that I do need to update the [README](https://github.com/robocode-dev/tank-royale/tree/master/bot-api/java#readme) with information on where to find the artifact on the local repository. :+1:

**Reply by SirStone** [reply]
Hello

I've just cloned the repo and launched a *gradle build*
I think the problem is still here. Looks very similar if not the same to me. I've tried all he above approaches without success and at the end I've undone everything and came back to the master HEAD, so the following error is from a clean cloned repo.

**here the build error output:**
```
~/R/tank-royale (master) [1]> gradle build

FAILURE: Build failed with an exception.

* Where:
Build file '~/Robocode-tank-royale/tank-royale/bot-api/java/build.gradle.kts'

* What went wrong:
Cannot get non-null property 'ossrhUsername' on project ':bot-api:java' as it does not exist
```

**my system:**
```
> gradle --version

------------------------------------------------------------
Gradle 7.4.2
------------------------------------------------------------

Build time:   2022-03-31 15:25:29 UTC
Revision:     540473b8118064efcc264694cbcaa4b677f61041

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          18.0.1 (Homebrew 18.0.1+0)
OS:           Mac OS X 10.15.7 x86_64
```

**Comment by flemming-n-larsen** [comment]
Thank you @SirStone for following up on this issue.

I made a fix so the `ossrhUsername` and `ossrhPassword` are now optional when these are missing in the _[user home]_/.gradle/gradle.properties file as these are only required when publishing the Bot API artifact to the Maven Central.

I have pushed the fix to the master branch.

**Reply by SirStone** [reply]
Thank you to follow up our issues so quickly. This project is probably taking a lot of your free time, I see passion here.
Back to the problem...I still see issues in building up from a clean and updated repo

The previous problem is gone, as you said, you fixed it. So the first *gradle build* try failed because was trying in building .Net stuff and I don't have it rugged up right now, so I commented these lines in *settings.gradle.kts*
- line 5 -> // include("schema:dotnet")
- line 18 -> // include("bot-api:dotnet")
- line 22 -> // include("sample-bots:csharp")
In order to exclude all the .Net products from the build process

Cleaning is all green, but I receive this error in building the project

```
~/R/tank-royale (master) [1]> gradle clean

BUILD SUCCESSFUL in 2s
9 actionable tasks: 9 up-to-date
~/R/tank-royale (master)> gradle build

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':server:shadowJar'.
> Could not resolve all dependencies for configuration ':server:runtimeClasspath'.
   > Could not find dev.robocode.tankroyale:robocode-tankroyale-schema:0.13.1.
     Searched in the following locations:
       - file:~/.m2/repository/dev/robocode/tankroyale/robocode-tankroyale-schema/0.13.1/robocode-tankroyale-schema-0.13.1.pom
       - https://repo.maven.apache.org/maven2/dev/robocode/tankroyale/robocode-tankroyale-schema/0.13.1/robocode-tankroyale-schema-0.13.1.pom
     Required by:
         project :server
```

Am I missing some steps? I followed the *https://repo.maven.apache.org/maven2/dev/robocode/tankroyale/robocode-tankroyale-schema/0.13.1/robocode-tankroyale-schema-0.13.1.pom* that the code is searching and I can't see it online, actually I can't open at all the *https://repo.maven.apache.org/maven2/dev/robocode/tankroyale/robocode-tankroyale-schema*; going back one slug in the url I can open *https://repo.maven.apache.org/maven2/dev/robocode/tankroyale* that shows a single folder for *robocode-tankroyale-bot-api* . This means that there are missing artifacts on Maven Central, at least, while I was trying this morning.

Thanks again for your efforts. With baby steps maybe one day I will start fixing things for you (an this community) instead of viceversa

Cheers

**Reply by flemming-n-larsen** [reply]
I realised that we did not need to create an artifact for the schema/java as the shadowJar task just uses the classes from the artifact. 

I have updated the master branch where there is no need to build a schema/jar artifact anymore. 🙂

**Reply by SirStone** [reply]
Hello
Thanks to holidays in Italy I had time to try again today. I've fixed my dotnet for sdk setup too and now, with no single modification to the code cloned (pulled) from the repo I'm able to do a full 'gradle build' with no errors
```
BUILD SUCCESSFUL in 3m 5s
43 actionable tasks: 40 executed, 3 up-to-date
```
Thanks for your effort, now maybe I can put some efforts of mine in this modifying the components for me or for the community

**Reply by flemming-n-larsen** [reply]
I am happy to learn that it is finally building as it should. I trying this out on Linux/Ubuntu and macOS from scratch to make sure it should work and fixed as many issues as possible.

Note:
When it comes to developing on the Bot APIs, one change to e.g. the Java version, will also need to be made for the .Net version and vice versa, so they stay in sync. At  a later point, more APIs should be added. For example a web based API, and then there will be at least 3 API's that needs to be in sync at that time.
